kafkalog/kafka.go

```diff
@@ -0,0 +1,351 @@
+				glog.Errorf("FAILED to send message: %s\n", err)
+				kafkaProducer.Close()
+				kafkaProducer = nil
+			}
+			pLock.Unlock()
+		}
+
+		if kafkaProducer != nil {
+			kafkaProducer.Close()
+		}
+
+		close(quit)
+	}()
+
+	return nil
+}
+
+type KafkaLogs struct {
+	Log    []byte `json:"log,omitempty"`
+	Stream string `json:"stream,omitempty"`
+
+	encoded []byte
+	err     error
+}
+
+func (d *KafkaLogs) ensureEncoded() {
+	if d.encoded == nil && d.err == nil {
+		d.encoded, d.err = json.Marshal(d)
+	}
+}
+
+func (d *KafkaLogs) Length() int {
+	d.ensureEncoded()
+	return len(d.encoded)
+}
+
+func (d *KafkaLogs) Encode() ([]byte, error) {
+	d.ensureEncoded()
+	return d.encoded, d.err
+}
+
+func newConsumer(hosts []string, container string, cfg logger.ReadConfig) (sarama.PartitionConsumer, int64, error) {
+	consumerLock.Lock()
+	defer consumerLock.Unlock()
+
+	tail := int64(cfg.Tail)
+
+	if err := initConsumerClient(hosts); err != nil {
+		glog.Errorf("Failed to initialize Sarama client: %v", err)
+		return nil, 0, err
+	}
+
+	newestOffset, err := kafkaConsumerCli.GetOffset(container, 0, sarama.OffsetNewest)
+	if err != nil {
+		glog.Errorf("Failed to get Sarama customer newest offset: %v", err)
+		return nil, 0, err
+	}
+
+	oldestOffset, err := kafkaConsumerCli.GetOffset(container, 0, sarama.OffsetOldest)
+	if err != nil {
+		glog.Errorf("Failed to get Sarama customer oldest offset: %v", err)
+		return nil, 0, err
+	}
+
+	glog.Infof("container %v, newest %v, oldest %v", container, newestOffset, oldestOffset)
+	if newestOffset == oldestOffset && !cfg.Follow {
+		return nil, 0, fmt.Errorf("no logs & no follow, return immediately")
+	}
+
+	consumer, err := sarama.NewConsumerFromClient(kafkaConsumerCli)
+	if err != nil {
+		glog.Errorf("Failed to create Sarama customer: %v", err)
+		return nil, 0, err
+	}
+
+	offset := sarama.OffsetOldest
+	if tail > 0 {
+		len := newestOffset - oldestOffset
+
+		if len >= tail {
+			offset = newestOffset - tail - 1
+		}
+	}
+
+	glog.Infof("tail %v, offset %v", tail, offset)
+
+	partitionConsumer, err := consumer.ConsumePartition(container, 0, offset)
+	if err != nil {
+		glog.Errorf("Failed to create Sarama customerPartition: %v", err)
+	}
+
+	return partitionConsumer, newestOffset, err
+}
+
+func New(ctx logger.Context) (logger.Logger, error) {
+	if _, ok := ctx.Config["KafkaHosts"]; !ok {
+		return nil, fmt.Errorf("Cannot find kafkaHosts in configuration")
+	}
+
+	return &KafkaLogger{
+		logs:      logChan,
+		hosts:     ctx.Config["KafkaHosts"],
+		container: ctx.ContainerID,
+		readers:   make(map[*logger.LogWatcher]struct{}),
+	}, nil
+}
+
+type KafkaLogger struct {
+	logs      chan *logger.Message
+	container string
+	hosts     string
+	mu        sync.Mutex
+	readers   map[*logger.LogWatcher]struct{} // stores the active log followers
+}
+
+func (k *KafkaLogger) Log(msg *logger.Message) error {
+	select {
+	case k.logs <- msg:
+	default:
+		return fmt.Errorf("log chan is full")
```

Comment:
@gao-feng OK, I agree block more. Then

if we hold the log here, we need to decide when should we drop the message, and we shouldn't hold it if the kafka is down. Will the consumer drain the `chan` if the connection is broken?

Other options:

- use longer queue, such as 4096 or longer. But I hate async log.
- use a kafka client pool, multiple clients write logs for different container, however, this won't help for a single crazy logging container.